### PR TITLE
executor: make the memory tracker of groupConcat more accurate.

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
We need track the cap of Buffer in partial result for groupConcat.

How it Works:

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: executor/aggfuncs/func_group_concat.go [hunk 1]
```diff
@@ -321,14 +341,14 @@ func (h *topNRows) tryToAdd(row sortRow) (truncated bool, sortRowMemSize int64)
 			h.rows[0].buffer.Truncate(h.rows[0].buffer.Len() - int(debt))
```
Comment (ichn-hu) [inline executor/aggfuncs/func_group_concat.go:341]: the truncate is also not tracked?

**Review by ichn-hu — COMMENTED** [review]
Rest LGTM

File: executor/aggfuncs/func_group_concat.go [hunk 2]
```diff
@@ -151,9 +163,11 @@ func (e *groupConcat) MergePartialResult(sctx sessionctx.Context, src, dst Parti
 		p2.buffer = p1.buffer
 		return 0, nil
 	}
+	memDelta -= int64(p2.buffer.Cap())
 	p2.buffer.WriteString(e.sep)
 	p2.buffer.WriteString(p1.buffer.String())
-	return 0, e.truncatePartialResultIfNeed(sctx, p2.buffer)
+	memDelta += int64(p2.buffer.Cap())
+	return memDelta, e.truncatePartialResultIfNeed(sctx, p2.buffer)
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/aggfuncs/func_group_concat.go:170]: No. trancate doesn't change the cap of the slice.

File: executor/aggfuncs/func_group_concat.go (hunk 1 above)
Comment (wshwsh12) [reply executor/aggfuncs/func_group_concat.go:341]: ditto

File: executor/aggfuncs/func_group_concat.go [hunk 3]
```diff
@@ -214,20 +236,19 @@ func (e *groupConcatDistinct) UpdatePartialResult(sctx sessionctx.Context, rowsI
...
 			p.buffer.WriteString(e.sep)
 		}
 		// write values
 		p.buffer.WriteString(p.valsBuf.String())
-		newMem := p.buffer.Cap()
-		memDelta += int64(newMem - oldMem)
 	}
+	memDelta += int64(p.valsBuf.Cap()) + int64(cap(p.encodeBytesBuffer))
 	if p.buffer != nil {
+		memDelta += int64(p.buffer.Cap())
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/aggfuncs/func_group_concat.go:251]: ditto...

**Review by ichn-hu — APPROVED** [review]

File: executor/aggfuncs/func_group_concat.go [hunk 4]
```diff
@@ -110,11 +115,19 @@ func (e *groupConcat) ResetPartialResult(pr PartialResult) {
...
+		memDelta += int64(-p.buffer.Cap())
+	}
 	for _, row := range rowsInGroup {
 		p.valsBuf.Reset()
 		for _, arg := range e.args {
 			v, isNull, err = arg.EvalString(sctx, row)
 			if err != nil {
+				memDelta += int64(p.valsBuf.Cap())
+				if p.buffer != nil {
+					memDelta += int64(p.buffer.Cap())
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_group_concat.go:129]: Why do we need to calculate the two `Cap` here?

File: executor/aggfuncs/func_group_concat.go [hunk 5]
```diff
@@ -321,14 +341,14 @@ func (h *topNRows) tryToAdd(row sortRow) (truncated bool, sortRowMemSize int64)
 			h.rows[0].buffer.Truncate(h.rows[0].buffer.Len() - int(debt))
 		} else {
 			h.currSize -= uint64(h.rows[0].buffer.Len()) + h.sepSize
+			memDelta -= int64(h.rows[0].buffer.Cap())
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_group_concat.go:344]: Why do we need to `-Cap` here?

File: executor/aggfuncs/func_group_concat.go [hunk 6]
```diff
@@ -500,12 +520,14 @@ func (e *groupConcatDistinctOrder) UpdatePartialResult(sctx sessionctx.Context,
...
 	p.topN.sctx = sctx
 	v, isNull := "", false
+	memDelta -= int64(cap(p.encodeBytesBuffer))
 	for _, row := range rowsInGroup {
 		buffer := new(bytes.Buffer)
 		p.encodeBytesBuffer = p.encodeBytesBuffer[:0]
 		for _, arg := range e.args {
 			v, isNull, err = arg.EvalString(sctx, row)
 			if err != nil {
+				memDelta += int64(cap(p.encodeBytesBuffer))
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_group_concat.go:530]: Can we defer this?

Comment (wshwsh12) [reply executor/aggfuncs/func_group_concat.go]: Yes. I have addressed. PTAL

File: executor/aggfuncs/func_group_concat.go (hunk 5 above)
Comment (wshwsh12) [reply executor/aggfuncs/func_group_concat.go:346]: The element has been pop from the heap. So the memory usage of the row should be minus.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**